common/service/apml/apml.c

```diff
@@ -0,0 +1,388 @@
+		return APML_ERROR;
+	}
+	return APML_SUCCESS;
+}
+
+static uint8_t access_RMI_mailbox(apml_msg *msg)
+{
+	int i = 0;
+
+	if (!check_mailbox_command_complete(msg, RETRY_MAX)) {
+		printf("[%s] previous command not complete.\n", __func__);
+		return APML_ERROR;
+	}
+
+	if (write_mailbox_request(msg)) {
+		printf("[%s] write request failed.\n", __func__);
+		return APML_ERROR;
+	}
+
+	/* wait for the requested command to complete */
+	if (!check_mailbox_command_complete(msg, RETRY_MAX)) {
+		printf("[%s] command not complete, retry %d times.\n", __func__, RETRY_MAX);
+		return APML_ERROR;
+	}
+
+	/* wait for SwAlertSts to be set */
+	uint8_t status;
+	for (i = 0; i < RETRY_MAX; i++) {
+		if (apml_read_byte(msg->bus, msg->target_addr, SBRMI_STATUS, &status)) {
+			printf("[%s] read SwAlertSts failed.\n", __func__);
+			return APML_ERROR;
+		}
+		if (status & 0x02) {
+			break;
+		}
+		k_msleep(WAIT_TIME_MS);
+	}
+	if (i == RETRY_MAX) {
+		printf("[%s] SwAlertSts not be set, retry %d times.\n", __func__, RETRY_MAX);
+		return APML_ERROR;
+	}
+
+	if (read_mailbox_response(msg)) {
+		printf("[%s] read mailbox response failed.\n", __func__);
+		return APML_ERROR;
+	}
+
+	/* clear SwAlertSts */
+	if (apml_write_byte(msg->bus, msg->target_addr, SBRMI_STATUS, 0x02)) {
+		printf("[%s] clear SwAlertSts failed.\n", __func__);
+		return APML_ERROR;
+	}
+	return APML_SUCCESS;
+}
+
+void apml_request_callback(apml_msg *msg)
+{
+	static uint8_t i = 0;
+	apml_resp_buffer[i].index = msg->ui32_arg & 0xFF;
+	memcpy(&apml_resp_buffer[i].msg, msg, sizeof(apml_msg));
+	i++;
+	if (i >= APML_RESP_BUFFER_SIZE) {
+		i = 0;
+	}
+}
+
+uint8_t get_apml_response_by_index(apml_msg *msg, uint8_t index)
+{
+	for (int i = 0; i < APML_RESP_BUFFER_SIZE; i++) {
+		if (apml_resp_buffer[i].index == index) {
+			memcpy(msg, &apml_resp_buffer[i].msg, sizeof(apml_msg));
+			return APML_SUCCESS;
+		}
+	}
+	return APML_ERROR;
+}
+
+uint8_t apml_read(apml_msg *msg)
+{
+	if (msg == NULL) {
+		printf("[%s] msg is NULL.\n", __func__);
+		return APML_ERROR;
+	}
+
+	if (k_msgq_put(&apml_msgq, msg, K_NO_WAIT)) {
+		printf("[%s] put msg to apml_msgq failed.\n", __func__);
+		return APML_ERROR;
+	}
+	return APML_SUCCESS;
+}
+
+static void apml_handler(void *arvg0, void *arvg1, void *arvg2)
+{
+	apml_msg msg_cfg;
+	while (1) {
+		memset(&msg_cfg, 0, sizeof(apml_msg));
+		k_msgq_get(&apml_msgq, &msg_cfg, K_FOREVER);
+
+		switch (msg_cfg.msg_type) {
+		case APML_MSG_TYPE_MAILBOX:
+			if (access_RMI_mailbox(&msg_cfg)) {
+				printf("[%s] access RMI mailbox failed.\n", __func__);
+				continue;
+			}
+			break;
+		case APML_MSG_TYPE_CPUID:
+			if (access_CPUID(&msg_cfg)) {
+				printf("[%s] access CPUID failed.\n", __func__);
+				continue;
+			}
+			break;
+		case APML_MSG_TYPE_MCA:
+			if (access_MCA(&msg_cfg)) {
+				printf("[%s] access MCA failed.\n", __func__);
+				continue;
+			}
+			break;
+		default:
+			break;
+		}
+
+		if (msg_cfg.cb_fn) {
+			msg_cfg.cb_fn(&msg_cfg);
+		}
+		k_msleep(10);
```

Comment:
It looks like this code is only executed when none of the other msg types are executed.
If that is true we should put it in the `default` case in the switch statement.

Reply:
It should be executed after accessing every msg_type except default.
I will modify code to achieve this.